### PR TITLE
pre-alpha impl of 1848

### PR DESCRIPTION
**Pull request by benjaminxscott**
Initial features re: #2548

Pulled most of this code from @eheiden's branch

Goal is to have a working baseline for people who want to implement this game's chrome/mechanics

Logos courtesy of @kelsin [at 18xx-maker](https://github.com/18xx-maker/18xx-maker/tree/master/src/data/logos/1848)

Open to a rebase / squash

### CODE REVIEW

File: assets/app/view/game/round/auction.rb [hunk 1]
```diff
@@ -170,6 +170,14 @@ def render_input(company)
               ]
             end
 
+          if @step.may_reduce?(company)
+            company_actions << h(:button, { on: { click: lambda {
+                                                           process_action(
```
Comment (benjaminxscott) [inline assets/app/view/game/round/auction.rb:175]: blame rubocop for this formatting

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          company.discount += 5
+          price = company.min_bid
+          @log << "#{current_entity.name} reduces #{company.name} by £5 to #{@game.format_currency(price)}"
+          @round.next_entity_index!
+        end
+
+        def may_reduce?(company)
+          # Each private can be discounted a maximum of 6 times
+
+          if company.sym == 'P1' && company.min_bid.positive?
```
Comment (benjaminxscott) [inline lib/engine/step/g_1848/dutch_auction.rb:70]: Probably a more generic way to handle this than hardcoding (e.g. "if discount > 30")
However, discount is already overloaded to handle the "buy-in price" differential, so I couldn't be bothered

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+          # P1 can be reduced to free, so we disable `check_positive`
+          player.spend(price, @game.bank, check_positive: false)
+          @companies.delete(company)
+          @log << "#{player.name} buys #{company.name} for #{@game.format_currency(price)}"
+
+          company.abilities(:share) do |ability|
+            share = ability.share
+
+            if share.president
+              # TODO: - I think this isn't necessary since we must par CAR to 100
```
Comment (benjaminxscott) [inline lib/engine/step/g_1848/dutch_auction.rb:138]: I started implementing the "hand the player a free share" similar to 18Ches, but ran into some errors

**Comment by perwestling** [comment]
I think we might need a check list for all the functionality of 1848, with those parts that have been implemented checked.
Create a new issue for that.

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 2 above)
Comment (jenf) [reply lib/engine/step/g_1848/dutch_auction.rb:70]: or a {'P2'=>40} hash and just look it up rather than multiple lines

Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb:70]: Good idea!

**Comment by benjaminxscott** [comment]
@perwestling agree

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 2 above)
Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb]: 410d102

File: lib/engine/action/reduce.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+module Engine
+  module Action
+    class Reduce < Base
+      attr_reader :company
+
+      def initialize(entity, company: nil)
+        @entity = entity
+        @company = company
+        raisenp 'Company cannot be nil' unless @company
```
Comment (tobymao) [inline lib/engine/action/reduce.rb:13]: raisenp? just raise a GameError

File: lib/engine/action/reduce.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
```
Comment (tobymao) [inline lib/engine/action/reduce.rb:1]: what is a reduce action

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        # On your turn, must buy any private or reduce price of any private
+        # Once you own a private, you may pass
+        # if everything is reduced to minimum price, and nobody has bought a private, you must buy one
+        # if any privates are left and everyone passes in a row, owned privates pay and auction continues
+        # if someone has bought a private and everything is minimum, you can pass
+
+        ACTIONS = %w[bid reduce].freeze
+        ACTIONS_WITH_PASS = %w[bid reduce pass].freeze
+
+        MIN_PRICES = {
```
Comment (tobymao) [inline lib/engine/step/g_1848/dutch_auction.rb:23]: why is this here? it's not the value of the company?

File: lib/engine/action/reduce.rb (hunk 5 above)
Comment (tobymao) [reply lib/engine/action/reduce.rb:1]: i feel like you can just use the assign action for this

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+          company.min_bid > MIN_PRICES[company.sym]
+        end
+
+        def actions(entity)
+          return [] if @companies.empty?
+          return [] unless entity.player?
+
+          actions = entity.player.companies.empty? ? ACTIONS : ACTIONS_WITH_PASS
+
+          entity == current_entity ? actions : []
```
Comment (tobymao) [inline lib/engine/step/g_1848/dutch_auction.rb:87]: why don't you do this current_entity check on line 83

File: lib/engine/action/reduce.rb (hunk 5 above)
Comment (benjaminxscott) [reply lib/engine/action/reduce.rb:1]: It "reduces the cost of a private"

File: assets/app/view/game/round/auction.rb (hunk 1 above)
Comment (tobymao) [reply assets/app/view/game/round/auction.rb:175]: put the hash on a new line

h(:button, {
  on: ...
})

File: lib/engine/action/reduce.rb (hunk 5 above)
Comment (tobymao) [reply lib/engine/action/reduce.rb:1]: yea, let's use that

File: lib/engine/step/base.rb [hunk 8]
```diff
@@ -36,6 +36,10 @@ def did_sell?(_corporation, _entity)
         false
       end
 
+      def may_reduce?(_company)
```
Comment (tobymao) [inline lib/engine/step/base.rb:39]: don't put this here, since no game uses it cept 48, just have it called in the view the action exists

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+        def process_pass(action)
+          entity = action.entity
+          @log << "#{entity.name} passes"
+          entity.pass!
+          all_passed! if entities.all?(&:passed?)
+          @round.next_entity_index!
+        end
+
+        def process_reduce(action)
```
Comment (tobymao) [inline lib/engine/step/g_1848/dutch_auction.rb:67]: you'll need to change this to process_assign

File: lib/engine/step/g_1848/dutch_auction.rb [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+          @game.payout_companies
+          @game.or_set_finished
+
+          entities.each(&:unpass!)
+        end
+
+        def buy_company(player, company, price)
+          company.owner = player
+          player.companies << company
+          # P1 can be reduced to free, so we disable `check_positive`
```
Comment (tobymao) [inline lib/engine/step/g_1848/dutch_auction.rb:123]: why don't you only do it when it's p1 then

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 6 above)
Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb:23]: Privates in this game are a bit weird, for instance

P1 has a "value" of 40 (can be bought in by a corp for up to 40)

P1 has a "price" of 30 (can be purchased by a player for 30)

P1 has a "minimum price" of zero (can be discounted by players taking "reduce" actions for up to 30$ total)

Other privates are similar, and each one can only be discounted by 30$ total

The `discount` data member is overloaded to also handle the difference between "price" and "value", so we can't just check whether discount >= 30 and so we have this hardcoded hash

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 7 above)
Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb:87]: I think this was copy/pasted from [the base `waterfall_auction`](https://github.com/tobymao/18xx/blob/master/lib/engine/step/waterfall_auction.rb#L64)

See 9b7e7b30

File: lib/engine/action/reduce.rb (hunk 4 above)
Comment (benjaminxscott) [reply lib/engine/action/reduce.rb:13]: 9b7e7b30

File: lib/engine/step/base.rb (hunk 8 above)
Comment (benjaminxscott) [reply lib/engine/step/base.rb]: 69c49307

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 10 above)
Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb:120]: You mean company.sym == 'P1' ... spend(..., check_positive=false)
else: spend(...)
?

File: assets/app/view/game/round/auction.rb (hunk 1 above)
Comment (benjaminxscott) [reply assets/app/view/game/round/auction.rb]: 57ed230f

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 10 above)
Comment (tobymao) [reply lib/engine/step/g_1848/dutch_auction.rb:120]: or spend(price, @game,bank, check_positive: company.sym != 'P1')

Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb]: 780735bd

File: lib/engine/action/reduce.rb (hunk 5 above)
Comment (benjaminxscott) [reply lib/engine/action/reduce.rb]: bca96f79017673bb5f853eb2899a76f486a7d7b8

File: lib/engine/step/g_1848/dutch_auction.rb (hunk 9 above)
Comment (benjaminxscott) [reply lib/engine/step/g_1848/dutch_auction.rb]: bca96f79017673bb5f853eb2899a76f486a7d7b8

**Comment by benjaminxscott** [comment]
I would def squashmerge this, since some of the early commits e.g. 411a056 were not relevant, and my later commit messages are not particularly meaningful

**Review by tobymao — APPROVED** [review]
